src/Components/Server/src/ComponentsHub.cs

```diff
@@ -48,15 +50,23 @@ internal CircuitHost CircuitHost
         /// <summary>
         /// Intended for framework use only. Applications should not call this method directly.
         /// </summary>
-        public override async Task OnDisconnectedAsync(Exception exception)
+        public override Task OnDisconnectedAsync(Exception exception)
```

Comment:
There's a fundamental concurrency problem here. We might need to have a registry of *all* circuits rather than just those that are **waiting*. 

If the client loses its connection and tries to reconnect before the service detects the disconnect, then it will spin up a new circuit. We basically need some way to know if a circuit is in use or not as well as knowing whether it's in the pool.